### PR TITLE
qdevices: modify the virtiofsd daemon '-d' as an optional parameter

### PR DESCRIPTION
**Pull request by zhenyzha**
Update virtiofsd daemon parameter to improve performance

ID: 1917638
Signed-off-by: Zhenyu Zhang <zhenyzha@redhat.com>

### CODE REVIEW

**Comment by zhenyzha** [comment]
On qemu4.2:
 (1/1) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.aarch64.io-github-autotest-qemu.virtio_fs_share_data.run_stress.with_xfstest.with_local_source.with_cache.auto.arm64-pci: FAIL: The xfstest failed. (1746.45 s)

@YongxueHong @menli820 pls help me to review the patch if you are free, thanks.

**Comment by YongxueHong** [comment]
Better to explain the detail about why to change it.

**Comment by zhenyzha** [comment]
> Better to explain the detail about why to change it.

ok, Thanks for the feedback.

About virtio-fs, in the initially developed framework, we added the -d parameter by default to start virtiofsd.

```
fsd_cmd += ' -d'
self.set_param('status_active', 'Waiting for vhost-user socket connection') 
```

This parameter will cause the ‘virtiofs daemon’ to print detailed thread call information during the entire startup process.
But in recent tests, we found that this parameter will significantly reduce the overall performance of virtio-fs.
So I suggest removing this default parameter from avocado-vt, 
If necessary for subsequent tests, we can add this parameter separately in tp-qemu.

**Comment by YongxueHong** [comment]
Yes, I got this reason about this PR,  but others didn't have idea about it.
So I suggest you write the details reason into the commit message.
Thanks.

**Comment by zhenyzha** [comment]
> Yes, I got this reason about this PR, but others didn't have idea about it.
> So I suggest you write the details reason into the commit message.
> Thanks.

ok
Regarding the code `self.set_param('status_active','Waiting for vhost-user socket connection')`,
test instructions:
In the original implementation of qemu4.2 virtio-fs, there is no output without the ‘-d’ parameter, so we added’-d’ to get the screen output “Waiting for vhost-user socket connection”
If the keyword status information is removed, the code will wait 1s before continuing by default.

```
if status_active:
    self._daemon_process.read_until_any_line_matches(
            status_active, timeout=read_until_timeout)
else:
    time.sleep(start_until_timeout)
```

After I deleted the code and tested it on qemu4.2, the logic can be executed normally. Wait for 1S after the virtiofs daemon command runs, and then start the guest.
So in avocado-vt, deleting the ‘-d’ parameter will not affect the original case.

**Comment by menli820** [comment]
LGTM.

related log:

01:26:15 INFO | Running virtiofs daemon command /usr/libexec/virtiofsd --socket-path=/var/tmp/avocado_pz0kjc22/avocado-vt-vm1-fs-virtiofsd.sock -o source=/var/lib/avocado/data/avocado-vt/virtio_fs_test/ -o cache=none.
01:26:17 INFO | Created virtiofs daemon process with parent PID 44892.
01:26:17 INFO | Running qemu command (reformatted):
MALLOC_PERTURB_=1  /usr/libexec/qemu-kvm \

**Comment by zhenyzha** [comment]
> LGTM.
> 
> related log:
> 
> 01:26:15 INFO | Running virtiofs daemon command /usr/libexec/virtiofsd --socket-path=/var/tmp/avocado_pz0kjc22/avocado-vt-vm1-fs-virtiofsd.sock -o source=/var/lib/avocado/data/avocado-vt/virtio_fs_test/ -o cache=none.
> 01:26:17 INFO | Created virtiofs daemon process with parent PID 44892.
> 01:26:17 INFO | Running qemu command (reformatted):
> MALLOC_PERTURB_=1 /usr/libexec/qemu-kvm \

@menli820 hi menghuan, Could you run virtio-fs case in qemu4.2 to prevent errors from appearing on the x86 platform?

**Comment by zhijianli88** [comment]
LGTM

**Review by YongxueHong — APPROVED** [review]

**Comment by menli820** [comment]
> > LGTM.
> > related log:
> > 01:26:15 INFO | Running virtiofs daemon command /usr/libexec/virtiofsd --socket-path=/var/tmp/avocado_pz0kjc22/avocado-vt-vm1-fs-virtiofsd.sock -o source=/var/lib/avocado/data/avocado-vt/virtio_fs_test/ -o cache=none.
> > 01:26:17 INFO | Created virtiofs daemon process with parent PID 44892.
> > 01:26:17 INFO | Running qemu command (reformatted):
> > MALLOC_PERTURB_=1 /usr/libexec/qemu-kvm \
> 
> @menli820 hi menghuan, Could you run virtio-fs case in qemu4.2 to prevent errors from appearing on the x86 platform?

run basic case in qemu4.2  on win2019 and rhel840 guest on x86 platform without issue, so the change on x86 platform is ok.

 (1/2) Host_RHEL.m8.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.none.q35: PASS (86.46 s)
 (2/2) Host_RHEL.m8.u3.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.none.q35: PASS (182.55 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

**Review by YongxueHong — APPROVED** [review]

File: virttest/qemu_devices/qdevices.py [hunk 1]
```diff
@@ -1877,15 +1877,13 @@ def start_daemon(self):
         fsd_cmd = '%s --socket-path=%s' % (self.get_param('binary'),
                                            self.get_param('sock_path'))
         fsd_cmd += ' -o source=%s' % self.get_param('source')
-        fsd_cmd += ' -d'
```
Comment (luckyh) [inline virttest/qemu_devices/qdevices.py:1880]: Suggest making whether using the daemon mode configurable.

Comment (YongxueHong) [reply virttest/qemu_devices/qdevices.py:1880]: Yes, make sense, Could we implement it through setting `self.get_param('extra_options')`, line 1882 ~ 1883 ?

**Comment by YongxueHong** [comment]
Hi @zhenyzha, @menli820 
Could you re-test it again with -d options to check whether there are issues?
you could modify the cfg, add `fs_extra_options = -d`, or add it to kar command line,
like, --customsparam="fs_extra_options = -d".
thanks.

File: virttest/qemu_devices/qdevices.py (hunk 1 above)
Comment (luckyh) [reply virttest/qemu_devices/qdevices.py:1880]: Looks possible, but just not that formal. Anyway, it's acceptable.

Comment (YongxueHong) [reply virttest/qemu_devices/qdevices.py:1880]: or could we check whether the call use `debug` mode then whether to execute the previous steps, for instance:
```
def __init__(self, aobject, binary, sock_path, source, extra_options=None, debug_mode=False):
    ...
    fsd_cmd += ' -o source=%s' % self.get_param('source')
    if debug_mode:
        fsd_cmd += ' -d'
    ...
    ...
    if debug_mode
        self.set_param('status_active', 'Waiting for vhost-user socket connection')
```

Comment (luckyh) [reply virttest/qemu_devices/qdevices.py:1880]: @YongxueHong yes, that would be an possible approach.

Comment (YongxueHong) [reply virttest/qemu_devices/qdevices.py:1880]: @luckyh 
Thanks, How about you? @zhenyzha

**Comment by zhenyzha** [comment]
> Hi @zhenyzha, @menli820
> Could you re-test it again with -d options to check whether there are issues?
> you could modify the cfg, add `fs_extra_options = -d`, or add it to kar command line,
> like, --customsparam="fs_extra_options = -d".
> thanks.

ok, update later

**Comment by zhenyzha** [comment]
> fs_extra_options

@YongxueHong Hi Can be added as expected.

```
- with_local_source:
     fs_binary_extra_options += " -d"
```

/usr/libexec/virtiofsd --socket-path=/tmp/avocado_qrvgi1kk/avocado-vt-vm1-fs1-virtiofsd.sock -o source=/tmp/virtio_fs1_test -o cache=auto -d
/usr/libexec/virtiofsd --socket-path=/tmp/avocado_qrvgi1kk/avocado-vt-vm1-fs2-virtiofsd.sock -o source=/tmp/virtio_fs2_test -o cache=auto -d

File: virttest/qemu_devices/qdevices.py (hunk 1 above)
Comment (zhenyzha) [reply virttest/qemu_devices/qdevices.py:1880]: @luckyh @YongxueHong It's ok for me, update later.

**Comment by zhenyzha** [comment]
@YongxueHong  pls help me to review the patch if you are free, thanks.

**Comment by menli820** [comment]
> Hi @zhenyzha, @menli820
> Could you re-test it again with -d options to check whether there are issues?
> you could modify the cfg, add `fs_extra_options = -d`, or add it to kar command line,
> like, --customsparam="fs_extra_options = -d".
> thanks.

fs_binary_extra_options += " -d"
It works as expected:
log:
06:01:46 INFO | Running virtiofs daemon command /usr/libexec/virtiofsd --socket-path=/tmp/avocado_h1htg9yi/avocado-vt-vm1-fs2-virtiofsd.sock -o source=/tmp/virtio_fs2_test -o cache=auto -d.

File: virttest/qemu_devices/qdevices.py [hunk 2]
```diff
@@ -1844,7 +1844,7 @@ class QVirtioFSDev(QDaemonDev):
     Virtiofs pseudo device.
     """
 
-    def __init__(self, aobject, binary, sock_path, source, extra_options=None):
+    def __init__(self, aobject, binary, sock_path, source, extra_options=None, debug_mode=False):
```
Comment (YongxueHong) [inline virttest/qemu_devices/qdevices.py:1847]: I have one question that how could I enable `debug_mode` in the case level?
From my side, I didn't find any interface to do that.
Thanks.

Comment (zhenyzha) [reply virttest/qemu_devices/qdevices.py:1847]: @YongxueHong The following are my test results, of course, if there is a better method, I am happy to improve.

```
with_local_source:
    debug_mode = yes

```

/usr/libexec/virtiofsd --socket-path=/tmp/avocado__doqcd1k/avocado-vt-vm1-fs1-virtiofsd.sock -o source=/tmp/virtio_fs1_test -d -o cache=auto

Comment (zhenyzha) [reply virttest/qemu_devices/qdevices.py:1847]: Thank you very much for your suggestions and support, update later.

**Comment by zhenyzha** [comment]
Test  with `fs_enable_debug_mode = yes` on tp-qemu:
`/usr/libexec/virtiofsd --socket-path=/tmp/avocado_iouy23mi/avocado-vt-vm1-fs1-virtiofsd.sock -o source=/tmp/virtio_fs1_test -d -o cache=auto`

Test  with `fs_enable_debug_mode = no`  or none fs_enable_debug_mode optional parameter on tp-qemu:
`/usr/libexec/virtiofsd --socket-path=/tmp/avocado_arhcufkx/avocado-vt-vm1-fs1-virtiofsd.sock -o source=/tmp/virtio_fs1_test -o cache=auto`

Interface function calls meet expectations.
@YongxueHong pls help me to review the patch if you are free, thanks.

**Comment by YongxueHong** [comment]
Hi @zhenyzha it looks good to me, 
Hi, @luckyh Do you have any comments?
Thanks.

**Review by luckyh — APPROVED** [review]

**Pull request merged**